File: ruby/tournament/tournament.rb
class Tournament
  FORMAT = "%-31s| %2s | %2s | %2s | %2s | %2s"
  def self.tally(input)
    h = {}
    input.split("\n") do |line|
      t1, t2, res = line.split(';')
      [t1, t2].each do |t|
        h[t] = Hash({ W:0, D:0, L:0, P:0, MP:0 }) unless h.key?(t)
        h[t][:MP] += 1
      end
      if res == 'win'
        h[t1][:W] += 1
        h[t1][:P] += 3
        h[t2][:L] += 1
      elsif res == 'loss'
        h[t2][:W] += 1
        h[t2][:P] += 3
        h[t1][:L] += 1
      elsif res == 'draw'
        h[t1][:D] += 1
        h[t1][:P] += 1
        h[t2][:D] += 1
        h[t2][:P] += 1
      end
    end
    out = FORMAT % %w[Team MP W D L P] + "\n"
    if not h.empty?
      (h.sort_by { |team, h| [-h[:P], team] }).each do |team, h|
        out += FORMAT % [team, h.slice(:MP,:W,:D,:L,:P).values] + "\n"
      end
    end
    out
  end
end

File: python/tournament/tournament.py
from collections import defaultdict
def tally(rows):
    tallies = defaultdict(int)
    for row in rows:
        teama, teamb, result = row.split(";")
        matches[teama]['matches'] += 1
        matches[teamb]['matches'] += 1
        if result == 'win':
            tallies[teama]['points'] += 3
            tallies[teama]['wins'] += 1
        elif result == 'loss':
            tallies[teamb]['points'] += 3
            tallies[teamb]['wins'] += 1
        elif result == 'draw':
            tallies[teama]['draws'] += 1
            tallies[teamb]['draws'] += 1
            tallies[teama]['points'] += 1
            tallies[teamb]['points'] += 1
    results = ["Team                           | MP |  W |  D |  L |  P"]
    string_ = ["|{:<20}|{:^4}||{:^4}|{:^4}|{:^4}|{:>4}"]
    for row in sorted(tallies.items(), key = lambda x: (-x['points'], x[0])):

        print(row)



print(tally(['Devastating Donkeys;Blithering Badgers;win', 'Devastating Donkeys;Blithering Badgers;win', 'Devastating Donkeys;Blithering Badgers;win', 'Devastating Donkeys;Blithering Badgers;win', 'Blithering Badgers;Devastating Donkeys;win']))
